### PR TITLE
feature: added netflix(via allflicks.net) support, disabled by default

### PR DESCRIPTION
**Pull request by geogolem**
configuration options are implemented under
Settings-->UI but their functionality is not complete
I still need to figure out how to make the configuration options work
properly. There are placeholders in the code
I simply need to be able to populate and retrieve these options and the
feature is complete

#### Database Migration
YES

#### Description
Added the ability for radarr to query allflicks.. Before this pull request is granted. I need some help completing the implementation of the 2 configuration options.

Please grep the changes for EnableAllFlicks/enableAllFlicks

and 

NetflixCountryCode/netflixCountryCode

I would appreciate any help to get those options working properly. Also any other feedback is appreciated.

I implemented a similar feature in CouchPotato and have been using it for several months. I really like the new Radarr, and will support this feature as required as best I can.

#### Todos
- [ ] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR

* Added ability for radarr to query allflicks.net
once this feature is complete, can implement a feature to disable downloading of movies known to be on netflix.

### CODE REVIEW

**Comment by geogolem** [comment]
i would like to have the feature disabled by default - -but i need some assistance implementing the configuration options...

**Comment by geogolem** [comment]
Added the ability for radarr to query allflicks..

**Comment by onedr0p** [comment]
I talked with @geogolem about this and he will be pushing further updates that will work nicer than sending a request to allflicks for the title.

**Comment by geogolem** [comment]
@onedr0p : Thanks for all your feedback and help. I checked in the working code... 

but i disabled the simplifiedcheck --> I didnt make an option for it yet:

  bool enableSimplifiedAllFlicksCheck = false; //_settingsService.EnableSimplifiedAllFLicksCheck;
                if (enableSimplifiedAllFlicksCheck)
                {
                    ;
                    //can add some code in here that will make use of for example: https://webduck.nl/plex2netflix/search?imdb=tt3171832
                    using (WebClient client = new WebClient())
                    {
                        string r1 = client.DownloadString("https://webduck.nl/plex2netflix/search?imdb=" + movie.ImdbId);
                        //Console.WriteLine(r1);
                        dynamic j1 = JObject.Parse(r1);
                        var results = j1.countries;
                        foreach (var result in results)
                        {
                            //Console.WriteLine(result);
                            if (result == countryCode)
                                movie.AllFlicksUrl = referer + "title/" + movie.TitleSlug;
                        }
                    }
                }
                else
xmlHTTPRequest to allflicks code

**Comment by geogolem** [comment]
@onedr0p : I tested out the simplified check and found one movie that got weird results..

Star Wars: Rogue One

for some reason https://webduck.nl/plex2netflix/search?imdb=tt3748528

returns: {"countries":["ca"],"checkedAt":1486113459}

this is incorrect. as Rogue One is not on Canadian netflix....

anyway this is just one example where it fails.. I need to do more testing/research to see what exactly is causing this. Rogue One is an easy example, because its  new movie and so I know its not on netflix... but the code needs to handle all the cases properly.  Look forward to your comments and following up tomorrow.

**Comment by geogolem** [comment]
@onedr0p : I also realize another solution is to screw the webduck.nl/plex2netflix check and just check to see if the allflicks url works out... I will experiment with this also..

Anyway, I wanted to potentially discuss my next steps with this feature.

I want to create another option like "Treat Netflix titles like they are downloaded locally"

In other words, when this feature is enabled --> Movies that are marked as being available in netflix will be treated as though they are already downloaded and in the library. They won't appear as missing or wanted, but rather will appear as though they have already been downloaded.

A few thoughts
Idea 1:
-------------------------
on the main radarr page we have: 
Blue: Missing, but not yet available. --> netflix has nothing to do with this category
Green: Downloaded and imported.  --> could have a category turqoise: Downloaded/imported but also on netflix
Red: Missing and monitored.
Orange: Missing, but not monitored.

Could have another category ---> yellow: Missing and on Netflix
-----------------------------

Idea 2:
In the left column  we have a picture of a ticket representing"In Cinemas"..
Maybe we could have the N (netflix logo) and "On Netflix" for titles on netflix (i.e. any title with the allflicks link will also be specified with this N logo

Then we could have an option "Treat 'On Netflix' similar to 'Downloaded and Imported'" and just represent these by some new color.

Keep in mind the Netflix status should not effect Monitoring status.. since an item can come and go in netflix at anytime... and if you were monitoring an item and it entered netflix.. it would not be considered missing but woud be considered similar to downloaded and imported.... but if the item then leaves netflix, it should be treated in a similar fashion to if a movie file was deleted from the harddisk; however, the monitor status should not change throughout, if the item was monitored prior to appearing on netflix it should still be monitored after it is gone from netflix etc.

Anyway these are just some of my thoughts..  I think after thinking about it.. I prefer Idea 2 more.. 

Basically have another category with another color called "Available on Netflix" or something like that... and treat it similar to Downloaded and Imported...

**Comment by geogolem** [comment]
after some more thought. I think I like the second idea better,

I think we could have Missing but on Netflix
Downloaded and On Netflix in the legend with some colour choices..

and then have another option On Netflix similar to the "Announced" or "In Cinemas" icons..

**Comment by geogolem** [comment]
icon-netflix-v6
can use something like that..

**Comment by geogolem** [comment]
i almost feel like "downloaded and imported" should just be called "collected"  ....  though its not really related.. i just had that thought... 

collected and on netflix
missing but on netflix

and all of them should basically be treated like they are collected....
released view should include on netflix items..

also unrelated... i noticed the table has a column for in cinemas... maybe this should be changed to initial release... because not all movies even go to cinemas.. some are straight to digital or tv movies... etc..

this is different than the icon.. im referrig to the column with the dates.

**Comment by geogolem** [comment]
The only thing that remains to be done in order to make these changes ready to be merged is the changing of the icon for netflix: 

UI/Content/icons.less:.icon-sonarr-movie-netflix {
UI/Movies/Editor/MovieEditorLayout.js:              icon     : 'icon-sonarr-movie-netflix',
UI/Movies/Index/MoviesIndexLayout.js:               icon     : 'icon-sonarr-movie-netflix',

and also the removal of the "On Netflix" filter if netflix is disabled in the UI options...

**Comment by geogolem** [comment]
I couldn't figure out how to observe a UI settings option in UI/Movies/Index/MoviesIndexLayout.js

File: src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs [hunk 1]
```diff
@@ -205,7 +210,92 @@ public Movie GetMovieInfo(int TmdbId, Profile profile = null)
...
+                    case "us":
+                    default:
+                        {
+                            referer = "https://www.allflicks.net/";
+                            break;
+                        }
+                }
+                if (enableAllFlicks=="simple")
+                {
+                    using (WebClient client = new WebClient())
```
Comment (galli-leo) [inline src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs]: Please use HttpClient everywhere.

File: src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs [hunk 2]
```diff
@@ -205,7 +210,92 @@ public Movie GetMovieInfo(int TmdbId, Profile profile = null)
...
+                        rquest.Method = "POST";
+                        rquest.Headers.Add("X-Requested-With", "XMLHttpRequest");
+                        rquest.Headers.Add("Cookie", cookIdent);
+                        rquest.ContentType = "application/x-www-form-urlencoded; charset=UTF-8";
+                        using (var st = rquest.GetRequestStream())
+                            st.Write(byteArray, 0, byteArray.Length);
+                        var rsponse = (HttpWebResponse)rquest.GetResponse();
+                        var rsponseString = new StreamReader(rsponse.GetResponseStream()).ReadToEnd();
+                        rsponse.Close();
+                        dynamic j1 = JObject.Parse(rsponseString);
```
Comment (galli-leo) [inline src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs]: Also please don't use dynamic objects. Create a class for each json object.

Comment (geogolem) [inline src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs]: i generated classes like this: public class Rootobject
{
    public int draw { get; set; }
    public int recordsTotal { get; set; }
    public int recordsFiltered { get; set; }
    public Datum[] data { get; set; }
}

public class Datum
{
    public string box_art { get; set; }
    public string title { get; set; }
    public string year { get; set; }
    public string genre { get; set; }
    public string rating { get; set; }
    public string available { get; set; }
    public string cast { get; set; }
    public string director { get; set; }
    public string genre2 { get; set; }
    public string genre3 { get; set; }
    public string slug { get; set; }
}

but .. wont this make the code more vulnerable to changes from allflicks.. i.e. if they added or removed a field etc?

File: src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs (hunk 1 above)
Comment (geogolem) [inline src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs]: I briefly looked into this. i havent been able to get it to work .. yet.. without simply using WebCLient inside: NzbDrone.Common/Http/HttpClient.cs

Comment (geogolem) [inline src/NzbDrone.Core/MetadataSource/SkyHook/SkyHookProxy.cs]: still looking into this.

**Comment by galli-leo** [comment]
Is this finished now?

**Comment by geogolem** [comment]
its finished and fully functional.. but I think I can improve it slightly.. so could we hold off on merging for a while longer on this...

**Comment by geogolem** [comment]
closing this pull request.. A new one with updated code and squashed commits will be created soon.

**Comment by geogolem** [comment]
it didnt actually die.. but i just sort of gave up bothering having it integrated into the actual build...

i do maintain a fork with a branch that has a much updated version of this with support for amazon prime video in addition to netflix as well as multiple countries/locales

i also managed to get the netflix and amazon logos in the filtering bar...

you can take a look at my branch here..

https://github.com/geogolem/Radarr/tree/JustWatchSupport

 if ppl try it out and want it integrated we can discuss that too...

**Comment by trevormiller6** [comment]
Thanks for the info. After looking at this a little more, this isn't exactly what i was looking for. It is awesome and something i would likely use though. 

I was looking for an option to not download movies that are netflix originals when adding movies from lists. Maybe during auto list import (netimport), check the hompage metadata from tmdb api before actually adding movie? idk C# so i am of no use, I got lost reading through the code.

something like:
```
if "netflix" in metadata["homepage"]:
    #dont import
    #call function to add movie to import exclusion
```

**Comment by geogolem** [comment]
Yes.. I see.. it's not the same as my feature.. in your case, you don't
even want the movie imported into radarr at all. My feature still allows
the movie to be imported but unmonitors titles on Netflix, not necessarily
only Netflix originals...

I'm curious.. do you definitely not want Netflix originals imported to
radarr.. or simply want to unmonitors Netflix originals only..

I would have to verify this tomorrow when I'm.in front of a PC.. but do all
Netflix originals absolutely have a home page with Netflix in the URL? If
that is the case it should be quite simple to implement what you want..
I'll look into it tomorrow ..

There are several possibilities. I'll write back here tomorrow with some
more details and then decide what is the best thing to implement.

Provided it is how I'm thinking it won't be too much effort to make
something like that work one way or another... I just need to look into it
more to see the best way to do it.

For example, to add to exclusions it might be better to have a more generic
and extendable feature that will cover your case but could be used for
others too.

**Comment by geogolem** [comment]
> Thanks for the info. After looking at this a little more, this isn't exactly what i was looking for. It is awesome and something i would likely use though.
> 
> I was looking for an option to not download movies that are netflix originals when adding movies from lists. Maybe during auto list import (netimport), check the hompage metadata from tmdb api before actually adding movie? idk C# so i am of no use, I got lost reading through the code.
> 
> something like:
> 
> ```
> if "netflix" in metadata["homepage"]:
>     #dont import
>     #call function to add movie to import exclusion
> ```

I looked into this.... but I need to clarify some things to better understand what we want to do here..
You are correct that the homepage is likely the best way to determine if a title is a netflix original or not... great. Lets just assume its trivial to determine if a title is a netflix original or not...

Import exclusions are only used when importing from a list... I dont think import exclusions is really useful here...  If you really only wanted to exclude from import it would be better to have it separate from import exclusions..

For example, an option "Exclude Netflix Originals" ... a tickbox in the settings just above or below the import exclusions in the radarr settings. If this box was not ticked everything functions as it does now. If it is ticked then it wont be imported during List Imports... so basically we dont touch import exclusions but treat these movies as though they were on the import exclusions list.... 

I dont think this would be too difficult to implement but I want to clarify since --- if a movie somehow is added to radarr i.e. manually or prior to using this feature etc... radarr will treat it like any other movie as the Netflix Original check would only be happening at list import... and not afterwards...

Another option, that is also easy to implement would be to add a "Unmonitor Netflix Originals" tickbox... During a Library update or an Update Movie Info task any netflix original would be unmonitored... Thus they would never be automatically downloaded.

Let me know your thoughts, Would you like one or both of these options? Am I overlooking something... ? Please let me know.

**Comment by baszalmstra** [comment]
I would very much like to have something like what your describing. Basically: "unmonitor everything I can watch on Netflix". I use an IMDB watchlists where I often add lots of movies that seem interesting. However, lots of them are available on Netflix these days so downloading them would be a complete waste.

I would still be very interested in this feature.

**Comment by geogolem** [comment]
> I would very much like to have something like what your describing. Basically: "unmonitor everything I can watch on Netflix". I use an IMDB watchlists where I often add lots of movies that seem interesting. However, lots of them are available on Netflix these days so downloading them would be a complete waste.
> 
> I would still be very interested in this feature.

Please try out the build here: https://github.com/geogolem/Radarr/tree/JustWatchSupport

You will have to enable the feature in the Settings --> UI tab...  Let me know if you have any problems/suggestions.

**Comment by trevormiller6** [comment]
> For example, an option "Exclude Netflix Originals" ... a tickbox in the settings just above or below the import exclusions in the radarr settings. If this box was not ticked everything functions as it does now. If it is ticked then it wont be imported during List Imports... so basically we dont touch import exclusions but treat these movies as though they were on the import exclusions list....
> 
> I dont think this would be too difficult to implement but I want to clarify since --- if a movie somehow is added to radarr i.e. manually or prior to using this feature etc... radarr will treat it like any other movie as the Netflix Original check would only be happening at list import... and not afterwards...
>

This is exactly what I am looking for! You are right that there is no need to add to import exclusion just treat them as if they were. The idea behind my request is that netflix originals most likely won't be removed from netflix ever so I will be able to watch them there always and i don't need to waste the storage. netflix does change the non-original movies they offer often so i would still like to download those movies when they pop up on one of the import lists. 

I appreciate your quick replies and willingness to work on this. sorry for the extremely late response; i suck.

**Comment by dantebarba** [comment]
What happened with this PR?

**Comment by bakerboy448** [comment]
> What happened with this PR? it was a great idea

Seems OP deleted their account

**Pull request closed without merging**